File: tests/test_qec.py
from __future__ import annotations

import math

import numpy as np
import pytest

import qupy as qp


def _small_model() -> qp.DetectorModel:
    return qp.DetectorModel(
        2,
        1,
        [
            qp.DetectorError(0.10, [0], [0]),
            qp.DetectorError(0.05, [1]),
            qp.DetectorError(0.01, [0, 1], [0]),
        ],
    )


def _large_model() -> qp.DetectorModel:
    errors: list[qp.DetectorError] = []
    for detector in range(32):
        observables = [0] if detector == 31 else []
        errors.append(qp.DetectorError(0.01 + 0.0001 * detector, [detector], observables))
    for detector in range(32):
        errors.append(
            qp.DetectorError(
                0.005 + 0.00005 * detector,
                [detector, (detector + 1) % 32],
            )
        )
    return qp.DetectorModel(32, 1, errors)


def _syndrome(model: qp.DetectorModel, correction: np.ndarray) -> np.ndarray:
    result = np.zeros(model.detector_count, dtype=np.int8)
    for selected, error in zip(correction, model.errors, strict=True):
        if selected:
            for detector in error.detectors:
                result[detector] ^= 1
    return result


def test_bp_osd_matches_exact_reference_on_small_model() -> None:
    model = _small_model()
    syndrome = np.array([1, 0], dtype=np.int8)

    exact = qp.decode_detector_model(model, syndrome)
    scalable = qp.decode_detector_model_bp_osd(model, syndrome)

    np.testing.assert_array_equal(_syndrome(model, scalable.correction), syndrome)
    np.testing.assert_array_equal(scalable.observables, exact.observables)
    assert scalable.log_likelihood == pytest.approx(exact.log_likelihood, abs=1e-12)
    assert scalable.matched_errors == exact.matched_errors
    assert scalable.method in {"belief-propagation", "belief-propagation-osd0"}
    assert not scalable.correction.flags.writeable
    assert not scalable.observables.flags.writeable


def test_bp_osd_zero_iterations_forces_order_zero_repair() -> None:
    decoder = qp.BpOsdDecoder(_small_model(), max_iterations=0)
    syndrome = np.array([1, 1], dtype=np.int8)
    result = decoder.decode(syndrome)

    assert result.osd_used
    assert not result.bp_converged
    assert result.iterations == 0
    assert result.method == "belief-propagation-osd0"
    np.testing.assert_array_equal(_syndrome(decoder.model, result.correction), syndrome)


def test_bp_osd_scales_beyond_reference_decoder_limit_and_batches() -> None:
    model = _large_model()
    known = np.zeros(len(model.errors), dtype=np.int8)
    known[[3, 17, 31]] = 1
    syndrome = _syndrome(model, known)

    with pytest.raises((ValueError, RuntimeError), match="at most 24"):
        qp.decode_detector_model(model, syndrome)

    decoder = qp.BpOsdDecoder(model, max_iterations=30, damping=0.1)
    result = decoder.decode(syndrome)
    np.testing.assert_array_equal(_syndrome(model, result.correction), syndrome)
    assert decoder.edge_count > 0
    assert decoder.active_error_count == 64
    assert math.isfinite(result.log_likelihood)

    syndromes = np.stack(
        [
            np.zeros(model.detector_count, dtype=np.int8),
            syndrome,
            syndrome,
        ]
    )
    batch = decoder.decode_batch(syndromes)

    assert batch.shots == 3
    assert batch.error_count == 64
    assert batch.observable_count == 1
    assert batch.corrections.shape == (3, 64)
    assert batch.observables.shape == (3, 1)
    assert batch.log_likelihoods.shape == (3,)
    assert batch.matched_errors.shape == (3,)
    assert batch.iterations.shape == (3,)
    assert batch.bp_converged.shape == (3,)
    assert batch.osd_used.shape == (3,)
    assert not batch.corrections.flags.writeable
    assert not batch.observables.flags.writeable
    np.testing.assert_array_equal(batch.corrections[1], result.correction)
    np.testing.assert_array_equal(batch.corrections[2], result.correction)
    for shot in range(batch.shots):
        np.testing.assert_array_equal(
            _syndrome(model, batch.corrections[shot]),
            syndromes[shot],
        )


def test_bp_osd_handles_deterministic_error_probabilities() -> None:
    model = qp.DetectorModel(
        2,
        1,
        [
            qp.DetectorError(1.0, [0], [0]),
            qp.DetectorError(0.2, [1]),
            qp.DetectorError(0.0, [0, 1], [0]),
        ],
    )
    result = qp.BpOsdDecoder(model, damping=0.25).decode([1, 1])

    np.testing.assert_array_equal(result.correction, [1, 1, 0])
    np.testing.assert_array_equal(result.observables, [1])
    np.testing.assert_array_equal(_syndrome(model, result.correction), [1, 1])


def test_bp_osd_rejects_impossible_syndrome_and_invalid_inputs() -> None:
    impossible = qp.DetectorModel(1, 0, [qp.DetectorError(0.0, [0])])
    with pytest.raises(ValueError, match="not representable"):
        qp.decode_detector_model_bp_osd(impossible, [1], max_iterations=0)

    with pytest.raises(ValueError, match="damping"):
        qp.BpOsdDecoder(_small_model(), damping=1.0)

    decoder = qp.BpOsdDecoder(_small_model())
    with pytest.raises(ValueError, match="one-dimensional"):
        decoder.decode([[0, 1]])
    with pytest.raises(ValueError, match="two-dimensional"):
        decoder.decode_batch([0, 1])
    with pytest.raises(ValueError, match="columns must match"):
        decoder.decode_batch(np.zeros((2, 3), dtype=np.int8))
    for invalid in ([0, 2], [0.5, 1.0], [0, 256]):
        with pytest.raises(ValueError, match="zero or one"):
            decoder.decode(invalid)
    with pytest.raises(ValueError, match="zero or one"):
        decoder.decode_batch([[0, 1], [1, 0.5]])

File: tests/cpp/test_qec.cpp
#include "qupy/qec.hpp"

#include <cmath>
#include <cstddef>
#include <cstdint>
#include <stdexcept>
#include <string>
#include <vector>

namespace {

void require(bool condition, const char* message) {
    if (!condition) {
        throw std::runtime_error(message);
    }
}

std::vector<std::int8_t> syndrome_from_correction(
    const qupy::DetectorModel& model,
    const std::vector<std::int8_t>& correction
) {
    require(correction.size() == model.errors().size(), "correction size mismatch");
    std::vector<std::int8_t> syndrome(model.detector_count(), 0);
    for (std::size_t error_index = 0U; error_index < correction.size(); ++error_index) {
        if (correction[error_index] == 0) {
            continue;
        }
        for (const std::size_t detector : model.errors()[error_index].detectors) {
            syndrome[detector] ^= 1;
        }
    }
    return syndrome;
}

qupy::DetectorModel small_model() {
    return qupy::DetectorModel(
        2U,
        1U,
        {
            qupy::DetectorError{0.10, {0U}, {0U}},
            qupy::DetectorError{0.05, {1U}, {}},
            qupy::DetectorError{0.01, {0U, 1U}, {0U}},
        }
    );
}

qupy::DetectorModel large_sparse_model() {
    std::vector<qupy::DetectorError> errors;
    errors.reserve(64U);
    for (std::size_t detector = 0U; detector < 32U; ++detector) {
        errors.push_back({
            0.01 + 0.0001 * static_cast<double>(detector),
            {detector},
            detector == 31U ? std::vector<std::size_t>{0U} : std::vector<std::size_t>{},
        });
    }
    for (std::size_t detector = 0U; detector < 32U; ++detector) {
        errors.push_back({
            0.005 + 0.00005 * static_cast<double>(detector),
            {detector, (detector + 1U) % 32U},
            {},
        });
    }
    return qupy::DetectorModel(32U, 1U, std::move(errors));
}

}  // namespace

int main() {
    {
        const qupy::DetectorModel model = small_model();
        const std::vector<std::int8_t> syndrome{1, 0};
        const qupy::DecodeResult exact = qupy::decode_detector_model(model, syndrome);
        const qupy::BpOsdDecodeResult scalable =
            qupy::decode_detector_model_bp_osd(model, syndrome, 20U, 0.0);
        require(
            syndrome_from_correction(model, scalable.correction) == syndrome,
            "small-model correction has wrong syndrome"
        );
        require(scalable.observables == exact.observables, "logical frame differs from exact");
        require(
            std::abs(scalable.log_likelihood - exact.log_likelihood) < 1e-12,
            "small-model likelihood differs from exact"
        );
        require(scalable.matched_errors == exact.matched_errors, "matched-error count differs");
    }

    {
        const qupy::DetectorModel model = small_model();
        const std::vector<std::int8_t> syndrome{1, 1};
        const qupy::BpOsdDecoder decoder(model, 0U, 0.0);
        const qupy::BpOsdDecodeResult result = decoder.decode(syndrome);
        require(result.osd_used, "zero-iteration decode should use OSD");
        require(!result.bp_converged, "zero-iteration decode unexpectedly reports BP convergence");
        require(result.iterations == 0U, "zero-iteration decode reports BP iterations");
        require(
            syndrome_from_correction(model, result.correction) == syndrome,
            "OSD correction has wrong syndrome"
        );
    }

    {
        const qupy::DetectorModel model(
            2U,
            1U,
            {
                qupy::DetectorError{1.0, {0U}, {0U}},
                qupy::DetectorError{0.2, {1U}, {}},
                qupy::DetectorError{0.0, {0U, 1U}, {0U}},
            }
        );
        const qupy::BpOsdDecoder decoder(model, 10U, 0.25);
        const qupy::BpOsdDecodeResult result = decoder.decode({1, 1});
        require(
            result.correction == std::vector<std::int8_t>({1, 1, 0}),
            "deterministic endpoint correction mismatch"
        );
        require(
            result.observables == std::vector<std::int8_t>({1}),
            "deterministic endpoint logical frame mismatch"
        );
        require(
            syndrome_from_correction(model, result.correction) ==
                std::vector<std::int8_t>({1, 1}),
            "deterministic endpoint syndrome mismatch"
        );
        require(decoder.active_error_count() == 1U, "wrong active-error count");
        require(decoder.edge_count() == 1U, "wrong Tanner edge count");
    }

    {
        const qupy::DetectorModel model(1U, 0U, {qupy::DetectorError{0.0, {0U}, {}}});
        bool rejected = false;
        try {
            (void) qupy::decode_detector_model_bp_osd(model, {1}, 0U, 0.0);
        } catch (const std::invalid_argument&) {
            rejected = true;
        }
        require(rejected, "impossible syndrome was accepted");
    }

    {
        bool rejected = false;
        try {
            (void) qupy::BpOsdDecoder(small_model(), 10U, 1.0);
        } catch (const std::invalid_argument&) {
            rejected = true;
        }
        require(rejected, "invalid damping was accepted");
    }

    {
        const qupy::DetectorModel model = large_sparse_model();
        require(model.errors().size() == 64U, "large model did not contain 64 errors");
        std::vector<std::int8_t> known(model.errors().size(), 0);
        known[3] = 1;
        known[17] = 1;
        known[31] = 1;
        const std::vector<std::int8_t> syndrome = syndrome_from_correction(model, known);

        bool exact_rejected = false;
        try {
            (void) qupy::decode_detector_model(model, syndrome);
        } catch (const std::length_error&) {
            exact_rejected = true;
        }
        require(exact_rejected, "reference decoder unexpectedly accepted 64 error mechanisms");

        const qupy::BpOsdDecoder decoder(model, 30U, 0.1);
        const qupy::BpOsdDecodeResult result = decoder.decode(syndrome);
        require(
            syndrome_from_correction(model, result.correction) == syndrome,
            "large-model correction has wrong syndrome"
        );
        require(result.correction.size() == model.errors().size(), "wrong correction width");
        require(
            result.observables.size() == model.observable_count(),
            "wrong logical-frame width"
        );

        const std::vector<std::int8_t> zero(model.detector_count(), 0);
        std::vector<std::int8_t> flattened;
        flattened.reserve(model.detector_count() * 3U);
        flattened.insert(flattened.end(), zero.begin(), zero.end());
        flattened.insert(flattened.end(), syndrome.begin(), syndrome.end());
        flattened.insert(flattened.end(), syndrome.begin(), syndrome.end());
        const qupy::BpOsdDecodeBatch batch = decoder.decode_batch(flattened, 3U);
        require(batch.shots == 3U, "wrong batch shot count");
        require(batch.error_count == model.errors().size(), "wrong batch error width");
        require(
            batch.observable_count == model.observable_count(),
            "wrong batch observable width"
        );
        require(
            batch.corrections.size() == 3U * model.errors().size(),
            "wrong batch correction storage"
        );
        require(
            batch.observables.size() == 3U * model.observable_count(),
            "wrong batch observable storage"
        );
        require(batch.log_likelihoods.size() == 3U, "wrong likelihood batch size");
        require(batch.matched_errors.size() == 3U, "wrong matched-error batch size");
        require(batch.iterations.size() == 3U, "wrong iteration batch size");
        require(batch.bp_converged.size() == 3U, "wrong convergence batch size");
        require(batch.osd_used.size() == 3U, "wrong OSD batch size");

        const std::vector<std::int8_t> batch_second(
            batch.corrections.begin() + static_cast<std::ptrdiff_t>(model.errors().size()),
            batch.corrections.begin() + static_cast<std::ptrdiff_t>(2U * model.errors().size())
        );
        require(batch_second == result.correction, "batch decode differs from single decode");
    }

    return 0;
}
